component/src/shielded_pool/component.rs

```diff
@@ -1,38 +1,51 @@
-use std::collections::BTreeSet;
+use std::collections::{BTreeMap, BTreeSet};
 
-use crate::{Component, Context};
+use crate::{
+    stake::{validator, View as _},
+    Component, Context,
+};
 use anyhow::{anyhow, Context as _, Result};
 use ark_ff::PrimeField;
 use async_trait::async_trait;
 use decaf377::{Fq, Fr};
 use penumbra_chain::{genesis, sync::CompactBlock, Epoch, KnownAssets, NoteSource, View as _};
 use penumbra_crypto::{
     asset::{self, Asset, Denom},
-    ka, note, Address, Note, NotePayload, Nullifier, One, Value, STAKING_TOKEN_ASSET_ID,
+    ka, note, Address, IdentityKey, Note, NotePayload, Nullifier, One, Value,
+    STAKING_TOKEN_ASSET_ID,
 };
+use penumbra_proto::Protobuf;
 use penumbra_storage::{State, StateExt};
 use penumbra_tct as tct;
-use penumbra_transaction::{Action, Transaction};
+use penumbra_transaction::{action::Undelegate, Action, Transaction};
 use tendermint::abci;
 use tracing::instrument;
 
-use crate::shielded_pool::{event, state_key, CommissionAmounts};
+use crate::shielded_pool::{event, state_key, CommissionAmounts, Quarantined};
 
-// Stub component
 pub struct ShieldedPool {
     state: State,
     note_commitment_tree: tct::Tree,
     /// The in-progress CompactBlock representation of the ShieldedPool changes
     compact_block: CompactBlock,
+    /// The scheduled quarantine actions for notes (processed all at once in end_block)
+    quarantined: Quarantined,
 }
 
 impl ShieldedPool {
     #[instrument(name = "shielded_pool", skip(state, note_commitment_tree))]
     pub async fn new(state: State, note_commitment_tree: tct::Tree) -> Self {
         Self {
-            state,
             note_commitment_tree,
-            compact_block: Default::default(),
+            // If a previous component started writing into the compact block (i.e. staking
+            // recording rolled back nullifiers and notes), retrieve it
+            compact_block: state
+                .compact_block(state.get_block_height().await.expect("can get height"))
+                .await
+                .expect("can get compact block")
+                .unwrap_or_default(),
```

Comment:
I don't understand how this can work -- if other components can write into the compact block, we can't keep it in memory.  This won't pick up any writes made by other components, because it's called when the components are being re-initialized, not after their writes.

Data sharing between components is *only* done through the `State`, so if we need other components to write into the `CompactBlock` (why?), we can't maintain an unsynchronized copy in memory.